File: 1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.java
class Solution {
    public int[] finalPrices(int[] prices) {
        int n=prices.length;
        int[] index = new int[prices.length];
 Stack<Integer> stack = new Stack<>();

 for(int i=n-1;i>=0;i--){
  if(stack.isEmpty()){
  index[i] = 0;
  }else if(prices[stack.peek()]>=prices[i]){
    while(!stack.isEmpty() && prices[stack.peek()]>prices[i]){
          stack.pop();
             }
            if(stack.isEmpty()){
            index[i] = 0;
             }else{
             index[i] = prices[stack.peek()];
            }
  }else{
  index[i] = prices[stack.peek()];
  }
  stack.push(i);
 }
        for(int i=0;i<n;i++){
            index[i]=prices[i]-index[i];
        }
 return index;
        
    }
}

File: 1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.py
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        tempstack =[]
        
        for i,price in enumerate(prices):
            while tempstack and prices[tempstack[-1]]>=price:
                index= tempstack.pop()
                prices[index] -=price
            tempstack.append(i)
        return prices
        